src/hotspot/share/runtime/threadSMR.cpp

```diff
@@ -847,6 +850,14 @@ bool ThreadsSMRSupport::delete_notify() {
   return (Atomic::load_acquire(&_delete_notify) != 0);
 }
 
+// Hash table size should be first power of two higher than twice the
+// length of the ThreadsList
```

Comment:
What is the basis for this? Seems reasonable for small numbers of threads but excessive for large numbers.

Reply:
This is where we need @fisk. This comment and the associated code was
written by @fisk in the early days of the Thread-SMR project. So we need
Erik to go spelunking in the recesses of his memory for why he sized things
this way...

Also, I think the comment is a bit wrong and doesn't match the code below.

Reply:
The theory in the code for KVHashtable or BasicHashtable is that it works better for prime numbers.  It'll resize to other prime numbers if needed.  I assume starting with 1007 should work fine, if the number of entries matches the number of threads.

const int _small_table_sizes[] = { 107, 1009, 2017, 4049, 5051, 10103, 20201, 40423 } ;
These are the resizing sizes in hashtable.cpp.

Reply:
> This is where we need @fisk. This comment and the associated code was
> 
> written by @fisk in the early days of the Thread-SMR project. So we need
> 
> Erik to go spelunking in the recesses of his memory for why he sized things
> 
> this way...
> 
> 
> 
> Also, I think the comment is a bit wrong and doesn't match the code below.

I have a habit of sizing hash tables in this way to achieve the following:
* The power of two table size means you can map to a bucket using hash & (size - 1), instead of the much more expensive hash % size. Modulo is several orders of magnitude more expensive to execute.
* Make sure the average number of entries per bucket is not exceeding 1.

Having said that, the first of those two reasons no longer applies; the more generic table logic we use nowadays uses modulo based bucket mapping I think. I probably had a more manually crafted table earlier where I used that property. It was quite a while ago, so I don't remember what happened with that. If I had to guess, I'd think @coleenp might have encouraged me to not implement yet another hash table, as we already have too many in general.